Criteria query in spring boot

Volodymyr-Mykychak opened the pull request.

Volodymyr-Mykychak committed:
✨ Implemented book search: added  endpoint, created , and updated

liliia-ponomarenko reviewed (changes requested):
Firstly, let’s fix CI/CD! Also, I see changes from the previous HW. Merge approved PR or change base branch.

liliia-ponomarenko commented on src/main/java/com/store/book/repository/book/BookSpecificationBuilder.java:
```diff
@@ -0,0 +1,41 @@
+        if (searchParametersDto.title() != null
+                && searchParametersDto.title().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager
+                    .getSpecificationProvider("title")
```
Make parameters like public constant fields and reuse them in providers

liliia-ponomarenko commented on src/main/java/com/store/book/repository/spec/AuthorSpecificationProvider.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public String getKey() {
+
```
```suggestion
```

liliia-ponomarenko commented on src/main/java/com/store/book/repository/spec/IsbnSpecificationProvider.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Override
+    public String getKey() {
+
```
```suggestion
```

liliia-ponomarenko reviewed (changes requested):
Good job! A few comments. 
Please note that your project is missing CI/CD and style checks. Without them, we cannot approve your solution.

Volodymyr-Mykychak committed:
🔧✨ Refactored specification builder to support multiple values
Updated BookSpecificationBuilder to handle String[] instead of String, ensuring compatibility with getSpecification(String[]). 🚀

JJJazl commented on src/main/java/com/store/book/repository/book/spec/AuthorSpecificationProvider.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Override
+    public String getKey() {
+
```
```suggestion
```

Elena-Bruyako replied:
remove empty line

JJJazl commented on src/main/java/com/store/book/repository/book/spec/AuthorSpecificationProvider.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Override
+    public Specification<Book> getSpecification(String[] params) {
+
```
```suggestion
```

Elena-Bruyako replied:
same here

JJJazl commented on src/main/java/com/store/book/repository/book/spec/IsbnSpecificationProvider.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public String getKey() {
+
```
Check all other places
```suggestion
```

Elena-Bruyako replied:
fix comment

JJJazl commented on src/main/java/com/store/book/repository/book/spec/AuthorSpecificationProvider.java:
```diff
@@ -0,0 +1,25 @@
+    public Specification<Book> getSpecification(String[] params) {
+
+        return (root, query, criteriaBuilder)
+                -> root.get("author").in(Arrays.stream(params).toArray());
```
Use `BookSpecificationField.AUTHOR.getKey()` constant. Same in other specification providers

Elena-Bruyako replied:
fix comment

JJJazl reviewed (changes requested):
Good! Please note that your project is missing CI/CD and style checks, let's fix it to move on)

Volodymyr-Mykychak committed:
add .github/workflows

Elena-Bruyako commented on src/main/java/com/store/book/repository/book/spec/IsbnSpecificationProvider.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public Specification<Book> getSpecification(String[] params) {
+
```
remove empty line
```suggestion
```

Elena-Bruyako commented on src/main/java/com/store/book/repository/book/spec/IsbnSpecificationProvider.java:
```diff
@@ -0,0 +1,26 @@
+    public Specification<Book> getSpecification(String[] params) {
+
+        return (root, query, criteriaBuilder)
+                -> root.get("isbn")
```
use constant that you already have

Elena-Bruyako commented on src/main/java/com/store/book/repository/book/spec/TitleSpecificationProvider.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public String getKey() {
+
```
same comments as in other classes

Elena-Bruyako reviewed (changes requested):
Please, fix previous comments

Volodymyr-Mykychak committed:
add constant to spec

Volodymyr-Mykychak committed:
add h2 data base

Elena-Bruyako commented on src/main/resources/application.properties:
```diff
@@ -1,8 +1,16 @@
-spring.jpa.show-sql=true
-spring.jpa.open-in-view=false
+
+# http://localhost:8080/h2-console
```
remove redundant empty lines

Elena-Bruyako commented on src/main/resources/application.properties:
```diff
@@ -1,8 +1,16 @@
-spring.datasource.username=user
-spring.datasource.password=acdACD123
-spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
+spring.datasource.url=jdbc:h2:mem:testdb
```
h2 should be used only for test
Please, revise [first HW](https://mate.academy/study/java/java-spring-boot/java-spring-boot-intro-v2?learnItemsFilter=All&section=practice&testTaskSlug=java_create_a_repo) in Spring module 

Pay attention to 

IMPORTANT section 

Create a new file application.properties in the src/test/resources folder
Add the following content to the test/application.properties file
```
spring.datasource.url=jdbc:h2:mem:testdb
spring.datasource.driverClassName=org.h2.Driver
spring.datasource.username=sa
spring.datasource.password=password
spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
```

Elena-Bruyako commented on data/bookstore.lock.db.trace.db:
```diff
@@ -0,0 +1,83 @@
+2025-02-28 13:04:52 jdbc[3]: exception
```
Remove data folder with files

Elena-Bruyako commented on src/main/java/com/store/book/repository/book/spec/AuthorSpecificationProvider.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public String getKey() {
+
```
There is still empty line (in other places also)

Elena-Bruyako reviewed (changes requested):
see comments

Volodymyr-Mykychak committed:
🔧 Fix configuration issues:
- 🚀 Move H2 settings to
- 🗑️ Remove redundant empty lines
- 💾 Ensure MySQL is used in production
- 🧹 Cleanup data files
- ✨ Remove unnecessary empty lines in code

Elena-Bruyako reviewed (changes requested):
Let's discussed your HW in chat (create new thread)

Volodymyr-Mykychak committed:
🔧 Fix configuration issues:
- 🚀 Move H2 settings to
- 🗑️ Remove redundant empty lines
- 💾 Ensure MySQL is used in production
- 🧹 Cleanup data files (bookstore.lock.db, trace.db)
- ✨ Remove unnecessary empty lines in AuthorSpecificationProvider.java

Elena-Bruyako reviewed (changes requested):
Let's discussed your HW in chat (create new thread)

Volodymyr-Mykychak committed:
🧹 Removed redundant empty lines and spaces in AuthorSpecificationProvider.java

Volodymyr-Mykychak committed:
🗑️ Removed changelog from test configuration to avoid conflicts

Volodymyr-Mykychak committed:
🔧 Updated test application.properties as per task requirements
